lib/ui/dart_runtime_hooks.cc

```diff
@@ -199,6 +202,42 @@ void Logger_PrintString(Dart_NativeArguments args) {
   }
 }
 
+void SaveCompilationTrace(Dart_NativeArguments args) {
+  const char* filePath = NULL;
+  {
+    Dart_Handle result = Dart_StringToCString(
+        Dart_GetNativeArgument(args, 0), &filePath);
+    if (Dart_IsError(result)) {
+      Dart_PropagateError(result);
+      return;
+    }
+  }
+
+  uint8_t* buffer = nullptr;
+  intptr_t size = 0;
+  Dart_Handle result = Dart_SaveCompilationTrace(&buffer, &size);
+  if (Dart_IsError(result)) {
+    FXL_LOG(ERROR) << "Failed to save trace: " << Dart_GetError(result);
+    return;
+  }
+
+  auto fd = fml::UniqueFD{
+      FML_HANDLE_EINTR(::open(filePath, O_CREAT|O_WRONLY|O_TRUNC, 0660))};
+  if (!fd.is_valid()) {
+    FXL_LOG(ERROR) << "Failed to create file: " << filePath << ", " << errno;
+    return;
+  }
+  if (write(fd.get(), buffer, (size_t) size) != (ssize_t) size) {
```

Comment:
Not that it matters since we probably shouldn't be using these file IO APIs here, but shouldn't you loop if size is less that what was requested for writes?

Reply:
I don't think I've seen this pattern. Do you have an example somewhere?